### PR TITLE
CSV based generator

### PR DESCRIPTION
**Pull request by yhenon**
Adds a new generator, based on ingesting a simple csv file. Useful feature for people to train on their own dataset!

### CODE REVIEW

**Comment by hgaiser** [comment]
Looks good, but I'm a bit concerned about how labels are created. Seems like it now depends on the order as they appear in the csv file? Perhaps it's better to sort them by name and then assign labels?

**Comment by yhenon** [comment]
I like that suggestion. Are you also considering a strategy to serialise
the labels to keep them consistent between train & test time?

**Comment by hgaiser** [comment]
Maybe it would make more sense to have the first row describe the classes? That way it can indeed be consistent. The method I proposed will also fail if there is not at least 1 annotation per object class in your CSV file.

**Comment by de-vri-es** [comment]
It's not exactly in the spirit of csv to have a separate list on the first line. You would normally only use it for a header for the remaining data, or not have it at all. But it seems the only other viable option would be a separate file. I'm not sure what is worse :)

**Comment by de-vri-es** [comment]
After thinking a bit more, I think that it does make sense to keep the class names in a separate file.

One reason is that the class name/number mapping is still useful without annotation. For example, a program running the network could use it to show human readable names.

Another reason is that it makes it really easy to combine and split CSV files with annotations. Tools like `cat` and `cut` will work perfectly. If we'd put class names on the first lines you would always need to treat the first line specially.

Note that it would be possible to add an option for automatically determining a mapping as an alternative to loading the mapping from a file.

File: keras_retinanet/preprocessing/csv_generator.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
+"""
+Copyright 2017-2018 Fizyr (https://fizyr.com)
```
Comment (de-vri-es) [inline keras_retinanet/preprocessing/csv_generator.py]: Thanks for the contribution, I really like the simplicity and re-usability of this generator.

We can't claim copyright for this though. In general you can not give away copyright without signing a legal document. So you should put your own copyright notice here.

If you developed this as part of your job (as employee) you should credit your employer. Otherwise, you can credit yourself (or a pseudonym if you don't want personal details in the file).

File: keras_retinanet/preprocessing/csv_generator.py [hunk 2]
```diff
@@ -1,6 +1,4 @@
 """
-Copyright 2017-2018 Fizyr (https://fizyr.com)
```
Comment (de-vri-es) [inline keras_retinanet/preprocessing/csv_generator.py]: Is there anything we can add as a copyright statement? Your real name, or otherwise your github name perhaps?

Comment (de-vri-es) [reply keras_retinanet/preprocessing/csv_generator.py]: Or should it be Netra Inc. ?

File: examples/train_csv.py [hunk 3]
```diff
@@ -39,8 +41,9 @@ def create_model(num_classes, weights='imagenet'):
 def parse_args():
     parser = argparse.ArgumentParser(description='Simple training script for object detection from a CSV file.')
     parser.add_argument('csv_path', help='Path to CSV file')
+    parser.add_argument('--classes', help='Path to a CSV file containing class label mapping.')
     parser.add_argument('--weights', help='Weights to use for initialization (defaults to ImageNet).', default='imagenet')
-    parser.add_argument('--batch-size', help='Size of the batches.', default=1, type=int)
+    parser.add_argument('--batch-size', help='Size of the batches.', default=2, type=int)
```
Comment (hgaiser) [inline examples/train_csv.py]: Why change the default to 2?

File: examples/train_csv.py [hunk 4]
```diff
@@ -60,19 +63,24 @@ def parse_args():
...
+    while True:
+        X, Y = train_generator.next()
 
     # create a generator for testing data
     test_generator = CSVGenerator(
-        args.csv_path,
-        'val',
-        test_image_data_generator,
+        csv_data_file=args.csv_path,
+        set_name='val',
```
Comment (hgaiser) [inline examples/train_csv.py]: Maybe make this an argument, defaulting to `val`.

File: keras_retinanet/preprocessing/csv_generator.py [hunk 5]
```diff
@@ -26,28 +28,46 @@
...
-        # csv with img_filepath, class, x1, y1, x2, y2, dataset
-        with open(csv_file, 'rb') as f_in:
+
+        if csv_class_file is not None:
+            #parse the provided class file
+            autogen_classes = False
+            self.classes = {}
+            with open(csv_class_file, 'rb') as f_class_in:
+                csvreader = csv.reader(f_class_in, delimiter=',')
+                for row in csvreader:
```
Comment (hgaiser) [inline keras_retinanet/preprocessing/csv_generator.py]: Maybe change this to `for classname, class_id in csvreader:`

File: keras_retinanet/preprocessing/csv_generator.py [hunk 6]
```diff
@@ -26,28 +28,46 @@
...
+                    classname, class_id = row
+                    self.classes[classname] = int(class_id)
+        else:
+            autogen_classes = True
+            self.classes              = []
```
Comment (hgaiser) [inline keras_retinanet/preprocessing/csv_generator.py]: Alignment

**Comment by yhenon** [comment]
Alright I'm fixing this up:
- there is now an optional `classname` argument in `examples/train_csv.py`. This can be used to pass a CSV file containing classname->id mappings. The parser will then respect this mapping, and throw an error if a class present in the dataset is missing in the mapping. For example the user could pass a CSV containing:
```
cat,0
dog,99
```
This would then create a 100 class model, with the required mappings.
- by default, the mappings are auto-generated by parsing the CSV. The classes are now sorted before being assigned an ID
- the copyright notice has been added
- some updates to the README to explain the use of the csv generator

File: keras_retinanet/preprocessing/csv_generator.py [hunk 7]
```diff
@@ -57,6 +77,13 @@ def __init__(
                     self.image_data[img_filepath] = []
                 self.image_data[img_filepath].append({'x1':int(x1), 'x2':int(x2), 'y1': int(y1), 'y2':int(y2), 'class':classname})
 
+        # sort the list of classes and convert to a dict
+
+        if autogen_classes:
+            self.classes = dict([(classname,ix) for ix, classname in enumerate(sorted(self.classes))])
+
+        print(self.classes)
```
Comment (hgaiser) [inline keras_retinanet/preprocessing/csv_generator.py]: I wouldn't print here.

File: keras_retinanet/preprocessing/csv_generator.py [hunk 8]
```diff
@@ -105,10 +132,10 @@ def load_annotations(self, image_index):
             box[0, 4] = self.name_to_label(class_name)
 
             bndbox = [annot['x1'], annot['y1'], annot['x2'], annot['y2']]
-            box[0, 0] = float(bndbox[0]) - 1
-            box[0, 1] = float(bndbox[1]) - 1
-            box[0, 2] = float(bndbox[2]) - 1
-            box[0, 3] = float(bndbox[3]) - 1
+            box[0, 0] = float(bndbox[0])-1
```
Comment (hgaiser) [inline keras_retinanet/preprocessing/csv_generator.py]: Why remove the spaces?

**Comment by yhenon** [comment]
> I was thinking if it makes more sense to have multiple files per set. So you'd have a train.txt CSV file for your training dataset and a test.txt file for your test dataset, etc. What do you think?

I agree. The user should now pass a CSV file for training, and, optionally a CSV file for validation.

In addition, passing a CSV file with classes is now required. This avoids difficulties associated with automatically inferring the classes for both the train and val file, especially if some classes are missing in one. The class file can be reused at inference time.

File: keras_retinanet/preprocessing/csv_generator.py [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+        return cv2.imread(path)
+
+    def load_annotations(self, image_index):
+        boxes = np.zeros((0, 5))
+
+        path  = self.image_names[image_index]
+        image = Image.open(path)
+
+        width = float(image.width)
+        height = float(image.height)
```
Comment (mateuszbuda) [inline keras_retinanet/preprocessing/csv_generator.py]: unused variables

File: keras_retinanet/preprocessing/csv_generator.py [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+
+            class_name = annot['class']
+
+            box[0, 4] = self.name_to_label(class_name)
+
+            bndbox = [annot['x1'], annot['y1'], annot['x2'], annot['y2']]
+            box[0, 0] = float(bndbox[0]) - 1
+            box[0, 1] = float(bndbox[1]) - 1
+            box[0, 2] = float(bndbox[2]) - 1
+            box[0, 3] = float(bndbox[3]) - 1
```
Comment (mateuszbuda) [inline keras_retinanet/preprocessing/csv_generator.py]: @yhenon Does this mean that you expect indexing of pixels in csv file to be 1 based (starting from 1) and here you transform it to 0 based, i.e. the first pixel has index 0?

File: keras_retinanet/preprocessing/csv_generator.py (hunk 9 above)
Comment (yhenon) [reply keras_retinanet/preprocessing/csv_generator.py]: unless I'm confused, you are looking at an outdated commit

File: keras_retinanet/preprocessing/csv_generator.py (hunk 10 above)
Comment (yhenon) [reply keras_retinanet/preprocessing/csv_generator.py]: this matches the behaviour in keras_retinanet/preprocessing/pascal_voc.py

I guess we can either: (a) define the csv format to index starting at 0 and remove the `-1` here, or (b) define the csv format to index starting at 1 and keep the `-1` here.

Comment (de-vri-es) [reply keras_retinanet/preprocessing/csv_generator.py]: Personally I feel we shouldn't copy weird choices made by pascal VOC. One-based indexing is odd. I'm in favor of dropping the -1 and defining the CSV format using zero based indices.

Comment (mateuszbuda) [reply keras_retinanet/preprocessing/csv_generator.py]: As long as it's documented somewhere, e.g. in readme, both options are good for me.

Comment (yhenon) [reply keras_retinanet/preprocessing/csv_generator.py]: fixed as per @de-vri-es comment which I agree with

**Comment by hgaiser** [comment]
Looks good to me. If @de-vri-es agrees I'd like  to merge this. I'm thinking to squash the commits to one commit too.
